### PR TITLE
Add UI for tracepoint callstack collection

### PR DESCRIPTION
**Pull request by Mahmood-Darwish**
This PR adds capture options that allow you to choose between different (if any) unwinding methods for the tracepoint call stack collection feature. The options are hidden behind the `tracepoint_callstack_collection` flag. Currently, choosing these options doesn't do anything as the feature is not completely implemented yet.

Bug: http://b/242681490

Screenshot: http://screenshot/8m9QmvkuVY5qJvE.png

### CODE REVIEW

File: src/OrbitQt/CaptureOptionsDialog.cpp [hunk 1]
```diff
@@ -194,6 +224,45 @@ DynamicInstrumentationMethod CaptureOptionsDialog::GetDynamicInstrumentationMeth
   ORBIT_UNREACHABLE();
 }
 
+void CaptureOptionsDialog::SetPureOrNotTracepoint(bool is_pure) {
```
Comment (beckerhe) [inline src/OrbitQt/CaptureOptionsDialog.cpp:227]: I have to admit `Pure` is a little confusing to me. May I suggest something more explicit?

What about `SetEnableCallStackCollectionOnThreadStateChanges`? (Ok, it's long. Maybe you have a better suggestion. :innocent: )

File: src/OrbitQt/CaptureOptionsDialog.cpp [hunk 2]
```diff
@@ -194,6 +224,45 @@ DynamicInstrumentationMethod CaptureOptionsDialog::GetDynamicInstrumentationMeth
...
+      ui_->tracepointCallstackDWARFMethodRadioButton->setChecked(true);
+      ui_->tracepointCallstackFramepointersMethodRadioButton->setChecked(false);
+      break;
+  }
+}
+
+orbit_grpc_protos::CaptureOptions::TracepointCallstackMethod
+CaptureOptionsDialog::GetTracepointCallstackMethod() const {
+  if (ui_->tracepointCallstackFramepointersMethodRadioButton->isChecked()) {
+    ORBIT_CHECK(!ui_->tracepointCallstackDWARFMethodRadioButton->isChecked());
```
Comment (beckerhe) [inline src/OrbitQt/CaptureOptionsDialog.cpp:256]: I don't think these CHECKs should be here. Nothing bad is gonna happen if they fail.

File: src/OrbitQt/CaptureOptionsDialog.cpp [hunk 3]
```diff
@@ -194,6 +224,45 @@ DynamicInstrumentationMethod CaptureOptionsDialog::GetDynamicInstrumentationMeth
...
 
+void CaptureOptionsDialog::SetPureOrNotTracepoint(bool is_pure) {
+  ui_->tracepointCallstackCollectionCheckBox->setChecked(!is_pure);
+}
+
+bool CaptureOptionsDialog::GetPureOrNotTracepoint() const {
+  return !ui_->tracepointCallstackCollectionCheckBox->isChecked();
+}
+
+void CaptureOptionsDialog::SetTracepointCallstackMethod(
```
Comment (beckerhe) [inline src/OrbitQt/CaptureOptionsDialog.cpp:235]: Nit: From a decoupling perspective I would suggest to make some name changes. The capture options dialog doesn't know that we use trace points for the call stack collection. It only knows that we offer the user to collect call stacks when a thread state change. And this happens to use Linux trace points. But it's not for the dialog to decide that we use trace points.

Imagine we were adding support for Windows in the future. Ideally we want to use the same option in the dialog, but the collection mechanism won't be tracepoints but ETW events.

(This is only a nit about naming. Feel free to ignore)

File: src/OrbitQt/CaptureOptionsDialog.cpp [hunk 4]
```diff
@@ -194,6 +224,45 @@ DynamicInstrumentationMethod CaptureOptionsDialog::GetDynamicInstrumentationMeth
...
+      ui_->tracepointCallstackDWARFMethodRadioButton->setChecked(false);
+      break;
+    case CaptureOptions::kDWARFTracepoint:
+      ui_->tracepointCallstackDWARFMethodRadioButton->setChecked(true);
+      ui_->tracepointCallstackFramepointersMethodRadioButton->setChecked(false);
+      break;
+    default:
+      ui_->tracepointCallstackDWARFMethodRadioButton->setChecked(true);
+      ui_->tracepointCallstackFramepointersMethodRadioButton->setChecked(false);
+      break;
```
Comment (beckerhe) [inline src/OrbitQt/CaptureOptionsDialog.cpp:249]: This is the same logic as for `DWARF`. Can you combine those two cases?

```cpp
case CaptureOptions::kDWARFTracepoint:
default:
// ...
```

should work

File: src/OrbitQt/orbitmainwindow.cpp [hunk 5]
```diff
@@ -1333,6 +1369,10 @@ void OrbitMainWindow::on_actionCaptureOptions_triggered() {
                     dialog.GetLimitLocalMarkerDepthPerCommandBuffer());
   settings.setValue(kMaxLocalMarkerDepthPerCommandBufferSettingsKey,
                     QString::number(dialog.GetMaxLocalMarkerDepthPerCommandBuffer()));
+  settings.setValue(kTracepointCallstackCollectionMethod,
+                    QString::number(dialog.GetTracepointCallstackMethod()));
+  settings.setValue(kIsPureTracepoint,
+                    QString::number(static_cast<int>(dialog.GetPureOrNotTracepoint())));
```
Comment (beckerhe) [inline src/OrbitQt/orbitmainwindow.cpp:1375]: Why `QString::number`? Just passing in ints should work. It does not?

**Review by beckerhe — APPROVED** [review]
Otherwise LGTM

**Review by beckerhe — APPROVED** [review]
Awesome. Thank you for all the changes. LGTM

File: src/GrpcProtos/capture.proto [hunk 6]
```diff
@@ -99,6 +99,15 @@ message CaptureOptions {
   repeated ApiFunction api_functions = 13;
 
   bool enable_api = 14;
+
+  enum TracepointCallstackMethod {
```
[the remark this answers is not in the document]
Comment (Mahmood-Darwish) [reply src/GrpcProtos/capture.proto]: Done.

File: src/OrbitQt/CaptureOptionsDialog.cpp [hunk 7]
```diff
@@ -6,6 +6,7 @@
 
 #include <absl/flags/declare.h>
 #include <absl/flags/flag.h>
+#include <qnamespace.h>
```
[the remark this answers is not in the document]
Comment (Mahmood-Darwish) [reply src/OrbitQt/CaptureOptionsDialog.cpp]: Done.

File: src/OrbitQt/CaptureOptionsDialog.cpp (hunk 1 above)
Comment (Mahmood-Darwish) [reply src/OrbitQt/CaptureOptionsDialog.cpp]: Done.

File: src/OrbitQt/CaptureOptionsDialog.cpp (hunk 2 above)
Comment (Mahmood-Darwish) [reply src/OrbitQt/CaptureOptionsDialog.cpp]: Done.

File: src/OrbitQt/CaptureOptionsDialog.cpp (hunk 3 above)
Comment (Mahmood-Darwish) [reply src/OrbitQt/CaptureOptionsDialog.cpp]: Dpme.

File: src/OrbitQt/CaptureOptionsDialog.cpp (hunk 4 above)
Comment (Mahmood-Darwish) [reply src/OrbitQt/CaptureOptionsDialog.cpp]: Done.

File: src/OrbitQt/orbitmainwindow.cpp (hunk 5 above)
Comment (Mahmood-Darwish) [reply src/OrbitQt/orbitmainwindow.cpp]: Done.

File: src/OrbitQt/CaptureOptionsDialog.cpp [hunk 8]
```diff
@@ -194,6 +225,40 @@ DynamicInstrumentationMethod CaptureOptionsDialog::GetDynamicInstrumentationMeth
...
+void CaptureOptionsDialog::SetThreadStateChangeCallstackMethod(
+    CaptureOptions::UnwindingMethod thread_state_change_callstack_method) {
+  switch (thread_state_change_callstack_method) {
+    case CaptureOptions::kFramePointers:
+      ui_->threadStateChangeCallstackFramepointersMethodRadioButton->setChecked(true);
+      ui_->threadStateChangeCallstackDWARFMethodRadioButton->setChecked(false);
+      break;
+    case CaptureOptions::kDwarf:
+    default:
+      ui_->threadStateChangeCallstackDWARFMethodRadioButton->setChecked(true);
```
Comment (karupayun) [inline src/OrbitQt/CaptureOptionsDialog.cpp:245]: Nit: Swap these lines to be consistent with the previous case.

```
 ui_->threadStateChangeCallstackFramepointersMethodRadioButton->setChecked(false);
 ui_->threadStateChangeCallstackDWARFMethodRadioButton->setChecked(true);
```

File: src/OrbitQt/CaptureOptionsDialog.cpp [hunk 9]
```diff
@@ -194,6 +225,40 @@ DynamicInstrumentationMethod CaptureOptionsDialog::GetDynamicInstrumentationMeth
...
+
+void CaptureOptionsDialog::SetThreadStateChangeCallstackMethod(
+    CaptureOptions::UnwindingMethod thread_state_change_callstack_method) {
+  switch (thread_state_change_callstack_method) {
+    case CaptureOptions::kFramePointers:
+      ui_->threadStateChangeCallstackFramepointersMethodRadioButton->setChecked(true);
+      ui_->threadStateChangeCallstackDWARFMethodRadioButton->setChecked(false);
+      break;
+    case CaptureOptions::kDwarf:
+    default:
```
Comment (karupayun) [inline src/OrbitQt/CaptureOptionsDialog.cpp:244]: Why the default here? I will move it below and log some error.

File: src/OrbitQt/CaptureOptionsDialog.cpp [hunk 10]
```diff
@@ -194,6 +225,40 @@ DynamicInstrumentationMethod CaptureOptionsDialog::GetDynamicInstrumentationMeth
...
+
+orbit_grpc_protos::CaptureOptions::UnwindingMethod
+CaptureOptionsDialog::GetThreadStateChangeCallstackMethod() const {
+  if (ui_->threadStateChangeCallstackFramepointersMethodRadioButton->isChecked()) {
+    return CaptureOptions::kFramePointers;
+  }
+  if (ui_->threadStateChangeCallstackDWARFMethodRadioButton->isChecked()) {
+    return CaptureOptions::kDwarf;
+  }
+  return CaptureOptions::kDwarf;
```
Comment (karupayun) [inline src/OrbitQt/CaptureOptionsDialog.cpp:259]: This is strange. I will rather return CaptureOptions::kUndefined;

File: src/OrbitQt/CaptureOptionsDialog.ui [hunk 11]
```diff
@@ -66,9 +66,9 @@
       <property name="geometry">
        <rect>
         <x>0</x>
-        <y>0</y>
+        <y>-435</y>
```
Comment (karupayun) [inline src/OrbitQt/CaptureOptionsDialog.ui:69]: This change is strange, but I guess it's right. Do you know why it could have happened?

File: src/OrbitQt/CaptureOptionsDialog.ui [hunk 12]
```diff
@@ -541,6 +541,48 @@ p, li { white-space: pre-wrap; }
             </property>
            </widget>
           </item>
+          <item>
+           <widget class="QCheckBox" name="threadStateChangeCallstackCollectionCheckBox">
```
Comment (karupayun) [inline src/OrbitQt/CaptureOptionsDialog.ui:545]: Consider adding a TODO to add a tooltip to explain to the users what these new options mean

File: src/OrbitQt/CaptureOptionsDialog.ui [hunk 13]
```diff
@@ -541,6 +541,48 @@ p, li { white-space: pre-wrap; }
...
+            </property>
+           </widget>
+          </item>
+          <item>
+           <widget class="QGroupBox" name="threadStateChangeCallstackUnwindingMethodGroupBox">
+            <property name="enabled">
+             <bool>true</bool>
+            </property>
+            <property name="title">
+             <string>Unwinding method</string>
```
Comment (karupayun) [inline src/OrbitQt/CaptureOptionsDialog.ui:560]: We've already had a Unwinding method group box, I found it a bit strange to have another one and that the options are not related at all. Consider explaining this by a tooltip as I mentioned before and also changing its name.

File: src/OrbitQt/orbitmainwindow.cpp [hunk 14]
```diff
@@ -1164,6 +1168,32 @@ void OrbitMainWindow::LoadCaptureOptionsIntoApp() {
...
+      static_cast<CaptureOptions::UnwindingMethod>(
+          settings
+              .value(kCallStackCollectionOnThreadStateChangesMethod,
+                     orbit_qt::CaptureOptionsDialog::kThreadStateChangeCallStackMethodDefaultValue)
+              .toInt());
+  if (settings.value(kCollectThreadStatesSettingKey, false).toBool()) {
+    if (!collect_callstack_on_thread_state_change) {
+      app_->SetThreadStateChangeCallstackCollection(
+          CaptureOptions::kNoThreadStateChangeCallStackCollection);
+      app_->SetThreadStateChangeCallstackMethod(orbit_grpc_protos::CaptureOptions::kUndefined);
```
Comment (karupayun) [inline src/OrbitQt/orbitmainwindow.cpp:1186]: I would expect that this is automatically set by the line above (setting to `kNoThreadStateChangeCallStackCollection`) or either that setting to Undefined is not needed anyways.

**Review by karupayun — COMMENTED** [review]
Hi Mahmood, nice PR! Everything is being wrapping up :) :)

I left a few comments!

File: src/OrbitQt/CaptureOptionsDialog.ui (hunk 11 above)
Comment (Mahmood-Darwish) [reply src/OrbitQt/CaptureOptionsDialog.ui:69]: I don't know why this is happening I edited everything through qt designer. I'm guessing this is correct since the menu shows up correctly and there's no need to change this right now.

File: src/OrbitQt/CaptureOptionsDialog.cpp (hunk 8 above)
Comment (Mahmood-Darwish) [reply src/OrbitQt/CaptureOptionsDialog.cpp]: Done.

File: src/OrbitQt/CaptureOptionsDialog.cpp (hunk 9 above)
Comment (Mahmood-Darwish) [reply src/OrbitQt/CaptureOptionsDialog.cpp]: Did I do it correctly now?

File: src/OrbitQt/CaptureOptionsDialog.cpp (hunk 10 above)
Comment (Mahmood-Darwish) [reply src/OrbitQt/CaptureOptionsDialog.cpp]: Done.

File: src/OrbitQt/CaptureOptionsDialog.ui (hunk 12 above)
Comment (Mahmood-Darwish) [reply src/OrbitQt/CaptureOptionsDialog.ui:545]: I added two tooltips.

File: src/OrbitQt/CaptureOptionsDialog.ui (hunk 13 above)
Comment (Mahmood-Darwish) [reply src/OrbitQt/CaptureOptionsDialog.ui]: Added.

File: src/OrbitQt/orbitmainwindow.cpp (hunk 14 above)
Comment (Mahmood-Darwish) [reply src/OrbitQt/orbitmainwindow.cpp:1186]: Setting to Undefined is not needed anyways but I did it anyway, it won't affect anything.

**Review by karupayun — APPROVED** [review]
LGTM! Thanks for addressing all my comments!
